ibis/mapd/client.py

```diff
@@ -0,0 +1,326 @@
+    dtypes = {
+        'BIGINT': dt.int64,
+        'BOOLEAN': dt.Boolean,
+        'CHAR': dt.string,
+        'DATE': dt.date,
+        'DECIMAL': dt.float64,
+        'DOUBLE': dt.float64,
+        'INT': dt.int32,
+        'INTEGER': dt.int32,
+        'FLOAT': dt.float32,
+        'NULL': dt.Null,
+        'NUMERIC': dt.float64,
+        'REAL': dt.float32,
+        'SMALLINT': dt.int16,
+        'STR': dt.string,
+        'TEXT': dt.string,
+        'TIME': dt.time,
+        'TIMESTAMP': dt.timestamp,
+        'VARCHAR': dt.string,
+    }
+
+    ibis_dtypes = {
+        v: k for k, v in dtypes.items()
+    }
+
+    def __init__(self, typename, nullable=False):
+        if typename not in self.dtypes:
+            raise com.UnsupportedBackendType(typename)
+        self.typename = typename
+        self.nullable = nullable
+
+    def __str__(self):
+        if self.nullable:
+            return 'Nullable({})'.format(self.typename)
+        else:
+            return self.typename
+
+    def __repr__(self):
+        return '<MapD {}>'.format(str(self))
+
+    @classmethod
+    def parse(cls, spec):
+        if spec.startswith('Nullable'):
+            return cls(spec[9:-1], nullable=True)
+        else:
+            return cls(spec)
+
+    def to_ibis(self):
+        return self.dtypes[self.typename](nullable=self.nullable)
+
+    @classmethod
+    def from_ibis(cls, dtype, nullable=None):
+        dtype_ = type(dtype)
+        if dtype_ in cls.ibis_dtypes:
+            typename = cls.ibis_dtypes[dtype_]
+        elif dtype in cls.ibis_dtypes:
+            typename = cls.ibis_dtypes[dtype]
+        else:
+            raise NotImplemented('{0} not Implemented'.format(dtype))
+
+        if nullable is None:
+            nullable = dtype.nullable
+        return cls(typename, nullable=nullable)
+
+
+class MapDCursor(object):
+    """Cursor to allow the MapD client to reuse machinery in ibis/client.py
+    """
+
+    def __init__(self, cursor):
+        self.cursor = cursor
+
+    def to_df(self):
+        if isinstance(self.cursor, Cursor):
+            col_names = [c.name for c in self.cursor.description]
+            result = pd.DataFrame(self.cursor.fetchall(), columns=col_names)
+        elif self.cursor is None:
+            result = pd.DataFrame([])
+        elif isinstance(self.cursor, pd.DataFrame):
+            result = self.cursor
+        elif GPUDataFrame is not None and isinstance(
+                self.cursor, GPUDataFrame
+        ):
+            result = self.cursor
+
+        return result
+
+    def __enter__(self):
+        # For compatibility when constructed from Query.execute()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        pass
+
+
+class MapDQuery(Query):
+    """
+
+    """
+    def _fetch(self, cursor):
+        # check if cursor is a pymapd cursor.Cursor
+        return self.schema().apply_to(cursor.to_df())
```

Comment:
We should probably come up with a simple interface for getting a `DataFrame` out of a `Cursor`, though you don't need to do that in this PR.

Reply:
That way you only have to inherit from `Query`.